crypto/rand/drbg_lib.c

```diff
@@ -0,0 +1,436 @@
+/*
+ * Copyright 2011-2017 The OpenSSL Project Authors. All Rights Reserved.
+ *
+ * Licensed under the OpenSSL license (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+#include <string.h>
+#include <openssl/crypto.h>
+#include <openssl/err.h>
+#include <openssl/rand.h>
+#include "rand_lcl.h"
+#include "rand_drbg_lcl.h"
+
+/*
+ * Support framework for SP800-90 DRBGs
+ */
+
+
+int drbg_fillin(DRBG_CTX *dctx, int type, unsigned int flags)
+{
+    int ret;
+
+    dctx->status = DRBG_STATUS_UNINITIALISED;
+    dctx->xflags = flags;
+    dctx->type = type;
+    dctx->iflags = 0;
+    dctx->entropy_blocklen = 0;
+    dctx->health_check_cnt = 0;
+    dctx->health_check_interval = DRBG_HEALTH_INTERVAL;
+
+    switch (type) {
+    default:
+        ret = -2;
+        break;
+    case NID_sha1:
+    case NID_sha224:
+    case NID_sha256:
+    case NID_sha384:
+    case NID_sha512:
+        ret = drbg_hash_init(dctx);
+        break;
+    case NID_aes_128_ctr:
+    case NID_aes_192_ctr:
+    case NID_aes_256_ctr:
+        ret = drbg_ctr_init(dctx);
+        break;
+    case NID_hmacWithSHA1:
+    case NID_hmacWithSHA224:
+    case NID_hmacWithSHA256:
+    case NID_hmacWithSHA512:
+        ret = drbg_hmac_init(dctx);
+        break;
+    }
+    if (ret == -2)
+        RANDerr(RAND_F_DRBG_FILLIN, RAND_R_UNSUPPORTED_DRBG_TYPE);
+    else if (ret < 0)
+        RANDerr(RAND_F_DRBG_FILLIN, RAND_R_ERROR_INITIALISING_DRBG);
+
+    return ret;
+}
+
+DRBG_CTX *RAND_DRBG_new(int type, unsigned int flags)
+{
+    DRBG_CTX *dctx = OPENSSL_zalloc(sizeof(*dctx));
+
+    if (dctx == NULL) {
+        RANDerr(RAND_F_RAND_DRBG_NEW, ERR_R_MALLOC_FAILURE);
+        return NULL;
+    }
+
+    if (drbg_fillin(dctx, type, flags) < 0) {
+        OPENSSL_free(dctx);
+        return NULL;
+    }
+    return dctx;
+}
+
+void RAND_DRBG_free(DRBG_CTX *dctx)
+{
+    if (dctx->uninstantiate)
+        dctx->uninstantiate(dctx);
+
+    /* Don't free up default DRBG */
+    if (dctx == RAND_DRBG_get_default()) {
+        memset(dctx, 0, sizeof(DRBG_CTX));
+        dctx->type = 0;
+        dctx->status = DRBG_STATUS_UNINITIALISED;
+    } else {
+        OPENSSL_cleanse(&dctx->d, sizeof(dctx->d));
+        OPENSSL_free(dctx);
+    }
+}
+
+static size_t drbg_get_entropy(DRBG_CTX *dctx, unsigned char **pout,
```

Comment:
The original fips_get_entropy() contained a bug which was reported by Lee D Gibbins <ldgibbons@avaya.com> on rt

	[rt.openssl.org #4055] FIPS Object Module User Guide corrections needed for (*get_entropy)()

(now GitHub Issue #2443) In contrary to the title, it was not an documentation issue but a real implementation bug, as it was confirmed by Steve Marquess in

 + https://mta.openssl.org/pipermail/openssl-dev/2015-October/002950.html
 
(The reply went only to [openssl-dev], not to [rt #4055] so it is missing in GitHub Issue #2443)

Since the bug could not be fixed, because the FIPS code was frozen, Steve Marquess just added a documentation of the bug to section 6.1.1 of the OpenSSL FIPS user guide on 2015-09-30:

	2015-09-30 Section 6.1.1, expanded discussion of the entropy callback (thanks to Lee D Gibbins ldgibbons@avaya.com) 

Now since the code now is being defrosted, this is the opportunity to finally fix this bug.